feat(label): refactored label component

mcoker opened the pull request:
fixes https://github.com/patternfly/patternfly/issues/2656

## Breaking changes
The label component has been redesigned and refactored. This PR introduces breaking visual and code changes. At a minimum to update to the new label component, the `.pf-c-label` component now requires `.pf-c-label__text` element to wrap the label text. And the label will be grey by default now, so to achieve the previous blue styling, the closest match visually would be using the new `.pf-m-blue` modifier. However, reference the new component documentation and the PR changeset for more info regarding the changes https://github.com/patternfly/patternfly/pull/2943/files

patternfly-build commented:
Preview: https://patternfly-pr-2943.surge.sh

maryshak1996 commented:
@mcoker these look great so far! 
@mcarrano to your quesiton about removable and clickable -- I think that they can live together just fine.
When the user hovers anywhere on the label that ISN'T the 'X', it would get the hover affect that I spelled out in my designs, and could click on the non-'X' areas for the clickable label functionality.
To dismiss / remove the label, they would hover over the 'X' itself, and the 'X' would go to the default icon hover color (black), and click on that 'X' to remove it

christiemolloy commented:
What do you think about the button being wider and less tall so that the focus ring fits within the constrain of the label, and you can use the extra width of the label to capture the area size

<img width="342" alt="Screen Shot 2020-04-13 at 2 08 23 PM" src="https://user-images.githubusercontent.com/20118816/79146352-4a16eb00-7d90-11ea-90fc-183b9b7118e0.png">

christiemolloy commented:
Is this hoverable or selectable ? when I think hoverable i think that i wouldn't be able to select it like I can here.

<img width="461" alt="Screen Shot 2020-04-13 at 2 10 28 PM" src="https://user-images.githubusercontent.com/20118816/79146561-94986780-7d90-11ea-982f-5c5064bf0304.png">

christiemolloy commented on src/patternfly/components/Label/examples/Label.md:
```diff
@@ -7,22 +7,553 @@ cssPrefix: pf-c-label
+| `.pf-m-compact` | `.pf-c-label` | Modifies label for compact styles. |
+| `.pf-m-outline` | `.pf-c-label` | Modifies label for outline styles. |
+| `.pf-m-hoverable` | `.pf-c-label` | Modifies label for a hoverable styles. |
+| `.pf-m-blue` | `.pf-c-label` | Modifies label for a blue styles. |
```
```suggestion
| `.pf-m-blue` | `.pf-c-label` | Modifies the label to have blue colored styling. |
```

christiemolloy commented on src/patternfly/components/Label/label.hbs:
```diff
@@ -3,4 +3,10 @@
     {{{label--attribute}}}
   {{/if}}>
   {{> @partial-block}}
+  {{#if label--IsClosable}}
```
Should this be `IsRemovable` ?

christiemolloy commented on src/patternfly/components/Label/label.scss:
```diff
@@ -1,31 +1,258 @@
 
-  display: inline-block;
+  // Hoverable
+  --pf-c-label--m-hoverable--hover--before--BorderWidth: var(--pf-global--BorderWidth--sm);
```
I left a comment but should this be called `selectable`

mcoker replied:
After talking to @mcarrano the use case for these is as links. So I'm thinking of styling `a` and `button` this way and adding a `.pf-m-clickable` variation  to be used with a non-interactive element like `<div class="pf-c-label">some text<button class="pf-c-button pf-m-plain">x</button</div>` to create a closable/clickable label since you can't put a `<button>` in a `<button>` or `<a>` element. wdyt?

christiemolloy replied:
ok that works!

mcoker replied:
Sorry so much has changed in the last day! Since @maryshak1996 prefers the hover outline not be shown on a link label when you hover the close button, I separated the label into 2 children - a `__content` element (that can be a `<span>` or `<a>`) and the close button. So you just use an `<a>` on `__content` if you want a link, and the close button is an adjacent element (instead of a child), so no need for the `clickable` class manipulation.

christiemolloy commented on src/patternfly/components/Label/examples/Label.md:
```diff
@@ -7,22 +7,553 @@ cssPrefix: pf-c-label
 
-{{#> label label--modifier="pf-m-compact"}}
-  Compact label
+{{#> label label--id="default-grey-icon"}}
```
did you include a compact variation in here?

maryshak1996 commented:
> What do you think about the button being wider and less tall so that the focus ring fits within the constrain of the label, and you can use the extra width of the label to capture the area size
> 
> <img alt="Screen Shot 2020-04-13 at 2 08 23 PM" width="342" src="https://user-images.githubusercontent.com/20118816/79146352-4a16eb00-7d90-11ea-90fc-183b9b7118e0.png">

@christiemolloy is that something that can be somewhat easily done? I wouldn't want to create extra work for you guys, if this alters how the icon area usually works

maryshak1996 commented:
sorry didn't mean to accidentally close this!

maryshak1996 commented:
> Is this hoverable or selectable ? when I think hoverable i think that i wouldn't be able to select it like I can here.
> 
> <img alt="Screen Shot 2020-04-13 at 2 10 28 PM" width="461" src="https://user-images.githubusercontent.com/20118816/79146561-94986780-7d90-11ea-982f-5c5064bf0304.png">

@christiemolloy on hover, the label should get an outline if it doesn't already have one. If there already is an outline, that outline should become 2px in thickness. I think that we should have the blue here on focus, but that shouldn't appear on hover

mcoker commented:
@maryshak1996 @christiemolloy re: the button, we can do whatever! I don't think we would want it to be any wider, unless we positioned the button farther to the right, otherwise the increased width would overlap the button text. We can make it shorter though, good call. I think the button is 2px longer on the top/bottom.

mcoker commented:
> When the user hovers anywhere on the label that ISN'T the 'X', it would get the hover affect that I spelled out in my designs, and could click on the non-'X' areas for the clickable label functionality.

@maryshak1996 I reworked the component so that a close button can be placed inside the label alongside the element that contains the text/image, and that element can be a link or button if the user chooses. I've updated the demo with more examples, including compact label examples. 

One change you'll see is that the text/icon now receives focus instead of the label itself, since the link/button has to be separate from the close button (a close button can't go inside of a link or another button). Let me know what you think of the focus behavior.

Since the label border changes serve as an indication that the label has focus, we could disable the default browser focus ring, too.

@christiemolloy addressed your comments

mcoker commented:
@maryshak1996 @mcarrano I've updated the grey/default label examples to be more robust to show all of the different states. Do you have a recommendation on how to present these examples, showing the options available while not having an overwhelming number of examples?

mcarrano commented:
Thanks for making all of these updates @mcoker . Here is some feedback.

* For the gray labels with icons, I think that the icon should be dark gray, not blue. Can you confirm @maryshak1996 ?
* Looks like you re-introduced the compact label? During design we had decided that there was not much benefit of having this. Do you still think that's the case @maryshak1996 ? I don't feel strongly about it, but if it is only minimal benefit then perhaps we should remove.
* I question whether we should have a label both as a link and a button. I understand the distinction, but not sure I see a use case for making a label be a button to invoke an action. Feel like having this might be misleading from a design perspective.

Good question about how best to present these examples. Maybe the first section in labeled "Variants" that shows all of the different combination with the gray filled labels. Then a second section that is "Label Styles"  and show all of the different color and border options, but I don't see a need to repeat all of the variants for both filled and outlined. Hopefully people will get that these are just different visual styles that can be applied to any label variant.

@mcoker @maryshak1996 what do you think?

maryshak1996 commented:
@mcarrano @mcoker 
I second what Matt said about the presentation of these examples (since there is a lot of variants, this seems like the most clear way to see it all). 
In regard to the color of the icon in the grey label, my initial thinking was the to have blue (to provide more contrast from the close icon), but @mcarrano I think that now, seeing these in context, having the icon be the black icon color is the way to go -- it does look out of place with it being blue. I'll make this update in the sketch library too. 
To the question about compact sizing -- ya the consensus was definitely to have just this one size (I mocked it up in more compact spaces and it still works just fine). I would be in favor of removing the 'compact size' for sure.

mcoker commented:
Great thanks @maryshak1996 @mcarrano! Removed the compact variation, and support for buttons. Also trimmed down the examples just to test all of the variations behave and look as expected. I created a new issue for the next sprint to clean up the examples - https://github.com/patternfly/patternfly/issues/2959

mcarrano commented:
I think this looks much better now @mcoker . You might not even need to refactor the examples as I think this looks OK now that you removed some of the variants. There is just one bug that I see. The "green removable link" is colored gray

![Screen Shot 2020-04-14 at 1 25 24 PM](https://user-images.githubusercontent.com/6619860/79255040-a42cb480-7e53-11ea-86ac-db7a3967e645.png)

mcoker commented:
@mcarrano thanks, squashed that bug.

maryshak1996 commented:
@mcoker this looks perfect to me!

christiemolloy commented on src/patternfly/components/Label/label.scss:
```diff
@@ -1,31 +1,271 @@
+    bottom: 0;
+    left: 0;
+    content: "";
+    border: var(--pf-c-label__content--before--BorderWidth) solid var(--pf-c-label__content--before--BorderColor);
```
why does this pseudo element need to be on `pf-c-label__content` ? can it just go on label?

mcoker replied:
The hover border treatment can't apply to the label since we don't want the label border to change when it's a link and the user hovers over the close button.

mcoker commented on src/patternfly/components/Label/label-content.hbs:
```diff
@@ -0,0 +1,9 @@
+<{{# if label-content--IsLink}}a{{else}}span{{/if}} class="pf-c-label__content{{#if label-content--modifier}} {{label-content--modifier}}{{/if}}"
```
Not necessarily. The `__content` element is the container for the icon and text, so we could just have the text inline with the icon. This gives us a container for the text so we have more control over it, though it's not necessary to have right now.

christiemolloy replied:
hmm.. i think the structure should be something similar to the button component. since we decided not to add the text wrapper, i don't know if its necessary to have two wrappers around the text in the label component ( `pf-c-label` pf-c-label__content` and `pf-c-label__text` ) @mattnolting thoughts?

mcoker replied:
works for me!

mcoker replied:
@christiemolloy actually just looked at it again and one reason we use the text element is, similar to chips, it serves as a value passed to the close button's `aria-labelledby` attribute.
